### ISSUE TITLE
(Not Sure) Stuck at `Executing database migration 'roomserver: published appservice pkey'` after upgraded to v0.11.1 from v0.11.0

### ISSUE THREAD

**Issue by bekcpear** [matrix-org/dendrite#2968]
### Background information

- **Dendrite version or git SHA**: `v0.11.1`
- **Monolith or Polylith?**: `Monolith`
- **SQLite3 or Postgres?**: `Postgres`
- **Running in Docker?**: **no**
- **`go version`**: `go version go1.20 linux/arm64`

### Description

- **What** is the problem:

After upgraded to v0.11.1, when I restart the service, it stucks at  `Executing database migration 'roomserver: published appservice pkey'` for a long time, and I don't know when it can finished.

```log
time="2023-02-10T19:38:18.013943932Z" level=info msg="Dendrite version 0.11.1"
time="2023-02-10T19:38:18.020023116Z" level=info msg="DNS cache enabled (size 256, lifetime 5m0s)"
time="2023-02-10T19:38:18.021271841Z" level=debug msg="Setting DB connection limits" conn_max_lifetime=-1s data_source_name="postgresql://127.0.0.1:3210/dendrite_db?sslmode=disable" max_idle_conns=5 max_open_conns=90
time="2023-02-10T19:38:18.046977022Z" level=debug msg="Using global database connection pool"
time="2023-02-10T19:38:18.065533334Z" level=debug msg="Executing database migration 'roomserver: published appservice pkey'"
```

I tried restarting a few times, but it didn't work.

**Comment by S7evinK** [comment]
Having seen your Gentoo repo, I'd guess you upgraded from 0.11.0?

**Comment by bekcpear** [comment]
> I'd guess you upgraded from 0.11.0?

yes

**Comment by devonh** [comment]
How long did you let it run for before restarting?
And how big is your `roomserver_published` table?

It could take a while to complete if the table is large.

**Comment by bekcpear** [comment]
> How long did you let it run for before restarting?

The longest is about 20 minutes

> And how big is your roomserver_published table?

I tried to use `pg_total_relation_size` to get but the db seems hanged up.

After back to normal, the size is:
```
dendrite_db=# SELECT pg_total_relation_size('roomserver_published');
 pg_total_relation_size 
------------------------
                  32768
(1 row)
```

---

It's weird the postgres is hanged up and even cannot be stop normally.

Because before restarting dendrite, the service is used normally, and the client can also be used normally, so I suspect that the database hang also occurs when restarting dendrite to use v0.11.1.

After force restarting postgres, dendrite starts back to normal.

However when I use element app to login, the session expired, and I have to verify the new login session. But NeoChat login is fine.

---

_I am now unable to determine whether this is caused by my problem, if so, please help close this Issue._
Thank you!

**Comment by bekcpear** [comment]
I checked the code diffs which effects database before `deltas.UpPulishedAppservicePrimaryKey` between v0.11.0 and v0.11.1 and find nothing weird. 

I also restored the backed database of v0.11.0 and test again under AMD64 platform with v0.11.1, and find everything is fine.

So I think it's an unexpected postgresql database hang event, although I can't understand why there is no problem with opening the database connection and the previous query statement.

I'm sorry to bother you, I'll close it.
